File: server_asp/src/api/user/UserController.cs
using Microsoft.AspNetCore.Mvc;
using shieldtify.common;
using shieldtify.middleware;
using shieldtify.models;
namespace shieldtify.api.user
{
    public static class UserController
    {
        [Tags("User -> Client")]
        public static APIRes getUser(HttpContext context)
        {
            try
            {
                return Middleware.MiddlewareAuthenticate(() =>
                {
                    var role = context.Items["Role"] as string;
                    if (role == "client")
                    {
                        var user = context.Items["User"] as ClientAccount;
                        return new APIRes(200, "Get user successfully", new
                        {
                            uid = user?.Uid,
                            username = user?.Username,
                            display_name = user?.DisplayName,
                            email = user?.Email,
                            role,
                            changed_password_at = user?.ChangedPasswordAt,
                            createdAt = user?.CreatedAt,
                            updatedAt = user?.UpdatedAt,
                        });
                    }
                    else
                    {
                        var user = context.Items["User"] as Account;
                        return new APIRes(200, "Get user successfully", new
                        {
                            uid = user?.Uid,
                            username = user?.Username,
                            display_name = user?.DisplayName,
                            role,
                            changed_password_at = user?.ChangedPasswordAt,
                            createdAt = user?.CreatedAt,
                            updatedAt = user?.UpdatedAt,
                        });
                    }
                }, context);
            }
            catch (Exception)
            {
                throw;
            }
        }
        [Tags("User -> Client")]
        public static APIRes getAddresses(HttpContext context)
        {
            try
            {
                return Middleware.MiddlewareAuthorize(() =>
                {
                    var user = context.Items["User"] as ClientAccount;
                    var DTO = UserService.getAddresses(user.Uid.ToString());
                    context.Response.StatusCode = DTO.statusCode;
                    return DTO;
                }, context, new List<string> { "client" });
            }
            catch (Exception)
            {
                throw;
            }
        }
        [Tags("User -> Client")]
        public static APIRes createAddress(HttpContext context, [FromBody] CreateAddressBody body)
        {
            try
            {
                return Middleware.MiddlewareAuthorize(() =>
                {
                    var user = context.Items["User"] as ClientAccount;
                    var DTO = UserService.createAddress(user.Uid.ToString(), body);
                    context.Response.StatusCode = DTO.statusCode;
                    return DTO;
                }, context, new List<string> { "client" });
            }
            catch (Exception)
            {
                throw;
            }
        }
        [Tags("User -> Admin")]
        public static APIRes getClients([FromQuery] string? page, string? keyword, HttpContext context)
        {
            try
            {
                return Middleware.MiddlewareAuthorize(() =>
                {
                    var DTO = UserService.getClients(
                        int.Parse(page ?? "1"),
                        keyword ?? ""
                    );
                    context.Response.StatusCode = DTO.statusCode;
                    return DTO;
                }, context, new List<string> { "admin", "superadmin" });
            }
            catch (Exception)
            {
                throw;
            }
        }
        [Tags("User -> Admin")]
        public static APIRes getClientById([FromRoute] string userId, HttpContext context)
        {
            try
            {
                return Middleware.MiddlewareAuthorize(() =>
                {
                    var DTO = UserService.getClientById(userId);
                    context.Response.StatusCode = DTO.statusCode;
                    return DTO;
                }, context, new List<string> { "admin", "superadmin" });
            }
            catch (Exception)
            {
                throw;
            }
        }
        [Tags("User -> Admin")]
        public static APIRes updateClient([FromRoute] string userId, [FromBody] UpdateClientBody body, HttpContext context)
        {
            try
            {
                return Middleware.MiddlewareAuthorize(() =>
                {
                    var DTO = UserService.updateClient(userId, body);
                    context.Response.StatusCode = DTO.statusCode;
                    return DTO;
                }, context, new List<string> { "admin", "superadmin" });
            }
            catch (Exception)
            {
                throw;
            }
        }

        [Tags("User -> Admin")]
        public static APIRes deleteAddressAdmin([FromRoute] string userId, [FromQuery] string addressId, HttpContext context)
        {
            try
            {
                return Middleware.MiddlewareAuthorize(() =>
                {
                    var DTO = UserService.deleteAddress(
                        userId,
                        addressId
                    );
                    context.Response.StatusCode = DTO.statusCode;
                    return DTO;
                }, context, new List<string> { "admin", "superadmin" });
            }
            catch (Exception)
            {
                throw;
            }
        }

        [Tags("User -> Admin")]
        public static APIRes getAccounts([FromQuery] string? page, string? keyword, HttpContext context)
        {
            try
            {
                return Middleware.MiddlewareAuthorize(() =>
                {
                    var DTO = UserService.getAccounts(
                        int.Parse(page ?? "1"),
                        keyword ?? ""
                    );
                    context.Response.StatusCode = DTO.statusCode;
                    return DTO;
                }, context, new List<string> { "admin", "superadmin" });
            }
            catch (Exception)
            {
                throw;
            }
        }

        [Tags("User -> Admin")]
        public static APIRes updateAccount([FromRoute] string id, [FromBody] UpdateAccountBody body, HttpContext context)
        {
            try
            {
                return Middleware.MiddlewareAuthorize(() =>
                {
                    var DTO = UserService.updateAccount(
                        context.Items["Role"] as string,
                        (context.Items["User"] as Account).Uid.ToString(),
                        id,
                        body
                    );
                    context.Response.StatusCode = DTO.statusCode;
                    return DTO;
                }, context, new List<string> { "admin", "superadmin" });
            }
            catch (Exception)
            {
                throw;
            }
        }

        [Tags("User -> Admin")]
        public static APIRes resetPassword([FromRoute] string id, HttpContext context)
        {
            try
            {
                return Middleware.MiddlewareAuthorize(() =>
                {
                    var DTO = UserService.resetPassword(
                        context.Items["Role"] as string,
                        id
                    );
                    context.Response.StatusCode = DTO.statusCode;
                    return DTO;
                }, context, new List<string> { "admin", "superadmin" });
            }
            catch (Exception)
            {
                throw;
            }
        }

        [Tags("User -> Client")]
        public static APIRes updateProfileClient([FromBody] UpdateClientBody body, HttpContext context)
        {
            try
            {
                return Middleware.MiddlewareAuthorize(() =>
                {
                    var DTO = UserService.updateClient(
                        (context.Items["User"] as ClientAccount).Uid.ToString(),
                        body
                    );
                    context.Response.StatusCode = DTO.statusCode;
                    return DTO;
                }, context, new List<string> { "client" });
            }
            catch (Exception)
            {
                throw;
            }
        }

        [Tags("User -> Admin")]
        public static APIRes updateAddressAdmin([FromRoute] string userId, [FromQuery] string addressId, [FromBody] CreateAddressBody body, HttpContext context)
        {
            try
            {
                return Middleware.MiddlewareAuthorize(() =>
                {
                    var DTO = UserService.updateAddress(
                        userId,
                        addressId,
                        body
                    );
                    context.Response.StatusCode = DTO.statusCode;
                    return DTO;
                }, context, new List<string> { "admin", "superadmin" });
            }
            catch (Exception)
            {
                throw;
            }
        }

        [Tags("User -> Client")]
        public static APIRes updateAddress([FromRoute] string addressId, [FromBody] CreateAddressBody body, HttpContext context)
        {
            try
            {
                return Middleware.MiddlewareAuthorize(() =>
                {
                    var DTO = UserService.updateAddress(
                        (context.Items["User"] as ClientAccount).Uid.ToString(),
                        addressId,
                        body
                    );
                    context.Response.StatusCode = DTO.statusCode;
                    return DTO;
                }, context, new List<string> { "client" });
            }
            catch (Exception)
            {
                throw;
            }
        }

        [Tags("User -> Client")]
        public static APIRes deleteAddressClient([FromRoute] string addressId, HttpContext context)
        {
            try
            {
                return Middleware.MiddlewareAuthorize(() =>
                {
                    var DTO = UserService.deleteAddress(
                        (context.Items["User"] as ClientAccount).Uid.ToString(),
                        addressId
                    );
                    context.Response.StatusCode = DTO.statusCode;
                    return DTO;
                }, context, new List<string> { "client" });
            }
            catch (Exception)
            {
                throw;
            }
        }

        [Tags("User -> Admin")]
        public static APIRes getAllOrders([FromQuery] string? page, string? keyword, HttpContext context)
        {
            try
            {
                return Middleware.MiddlewareAuthorize(() =>
                {
                    var DTO = UserService.getAllOrders(
                        int.Parse(page ?? "1"),
                        keyword ?? ""
                    );
                    context.Response.StatusCode = DTO.statusCode;
                    return DTO;
                }, context, new List<string> { "admin", "superadmin" });
            }
            catch (Exception)
            {
                throw;
            }
        }

        [Tags("User -> Client")]
        public static APIRes getOrderByClientID(HttpContext context)
        {
            try
            {
                return Middleware.MiddlewareAuthorize(() =>
                {
                    var DTO = UserService.getOrderByClientID(
                        (context.Items["User"] as ClientAccount).Uid.ToString()
                    );
                    context.Response.StatusCode = DTO.statusCode;
                    return DTO;
                }, context, new List<string> { "client" });
            }
            catch (Exception)
            {
                throw;
            }
        }

        [Tags("User -> Client")]
        public static APIRes getOrderByIdClient([FromRoute] string orderId, HttpContext context)
        {
            try
            {
                return Middleware.MiddlewareAuthorize(() =>
                {
                    var DTO = UserService.getOrderByID(
                        orderId,
                        (context.Items["User"] as ClientAccount).Uid.ToString()
                    );
                    context.Response.StatusCode = DTO.statusCode;
                    return DTO;
                }, context, new List<string> { "client" });
            }
            catch (Exception)
            {
                throw;
            }
        }
        [Tags("User -> Admin")]
        public static APIRes getOrderByIdAdmin([FromQuery] string userId, [FromQuery] string orderId, HttpContext context)
        {
            try
            {
                return Middleware.MiddlewareAuthorize(() =>
                {
                    var DTO = UserService.getOrderByID(
                        orderId,
                        userId
                    );
                    context.Response.StatusCode = DTO.statusCode;
                    return DTO;
                }, context, new List<string> { "admin", "superadmin" });
            }
            catch (Exception)
            {
                throw;
            }
        }
        [Tags("User -> Admin")]
        public static APIRes createStaff([FromBody] CreateStaffBody body, HttpContext context)
        {
            try
            {
                return Middleware.MiddlewareAuthorize(() =>
                {
                    var DTO = UserService.createStaff(
                        context.Items["Role"] as string,
                        body
                    );
                    context.Response.StatusCode = DTO.statusCode;
                    return DTO;
                }, context, new List<string> { "superadmin", "admin" });
            }
            catch (Exception)
            {
                throw;
            }
        }
        [Tags("User -> Admin")]
        public static APIRes getOrdersByStatus([FromQuery] string status, string? page, string? keyword, HttpContext context)
        {
            try
            {
                return Middleware.MiddlewareAuthorize(() =>
                {
                    var DTO = UserService.getOrdersByStatus(
                        status,
                        int.Parse(page ?? "1"),
                        keyword ?? ""
                    );
                    context.Response.StatusCode = DTO.statusCode;
                    return DTO;
                }, context, new List<string> { "superadmin", "admin", "staff" });
            }
            catch (Exception)
            {
                throw;
            }
        }
        [Tags("User -> Admin")]
        public static APIRes updateOrder([FromRoute] string orderId, [FromBody] UpdateOrderBody body, HttpContext context)
        {
            try
            {
                return Middleware.MiddlewareAuthorize(() =>
                {
                    var DTO = UserService.updateOrder(
                        (context.Items["User"] as Account).Uid.ToString(),
                        orderId,
                        body
                    );
                    context.Response.StatusCode = DTO.statusCode;
                    return DTO;
                }, context, new List<string> { "superadmin", "admin", "staff" });
            }
            catch (Exception)
            {
                throw;
            }
        }
        [Tags("User -> Admin")]
        public static APIRes processOrders([FromQuery] string? type, [FromBody] List<string> orders, HttpContext context)
        {
            try
            {
                return Middleware.MiddlewareAuthorize(() =>
                {
                    var DTO = UserService.processOrders(
                        (context.Items["User"] as Account).Uid.ToString(),
                        orders,
                        int.Parse(type ?? "1")
                    );
                    context.Response.StatusCode = DTO.statusCode;
                    return DTO;
                }, context, new List<string> { "superadmin", "admin", "staff" });
            }
            catch (Exception)
            {
                throw;
            }
        }
    }

    public class CreateAddressBody
    {
        public required string address { get; set; }
        public required string city { get; set; }
        public required string province { get; set; }
        public required string phone_number { get; set; }
        public required bool is_default { get; set; }
    }

    public class UpdateClientBody
    {
        public string? username { get; set; }
        public string? display_name { get; set; }
    }

    public class UpdateAccountBody
    {
        public string? username { get; set; }
        public string? display_name { get; set; }
        public string? role { get; set; }
    }

    public class UpdateOrderBody
    {
        public string uid { get; set; }
        public string? order_status { get; set; }
        public string? shipping_addressid { get; set; }
        public string? payment_method { get; set; }
        public string? receive_method { get; set; }
        public List<OrderItemBody>? products { get; set; }
    }

    public class OrderItemBody
    {
        public string itemid { get; set; }
        public int? quantity { get; set; }
        public float? new_price { get; set; }
    }

    public class CreateStaffBody
    {
        public string username { get; set; }
        public string display_name { get; set; }
        public string role { get; set; }
    }
}

File: server_nodejs/src/api/user/user.controller.js
const service = require('./user.service');

module.exports = {
    getUser: async (req, res, next) => {
        try {
            res.status(200).json({
                statusCode: 200,
                message: 'Get user successfully',
                data: req.user,
            });
        } catch (error) {
            next(error);
        }
    },
    getAdresses: async (req, res, next) => {
        try {
            const user = req.user;
            const DTO = await service.getAddresses(user.uid);
            res.status(DTO.statusCode).json(DTO);
        } catch (error) {
            next(error);
        }
    },
    createAddress: async (req, res, next) => {
        try {
            const user = req.user;
            const DTO = await service.createAddress(user.uid, req.body);
            res.status(DTO.statusCode).json(DTO);
        } catch (error) {
            next(error);
        }
    },
    getClients: async (req, res, next) => {
        try {
            const DTO = await service.getClients(
                req.query.page,
                req.query.keyword,
            );
            res.status(DTO.statusCode).json(DTO);
        } catch (error) {
            next(error);
        }
    },
    getClientById: async (req, res, next) => {
        try {
            const DTO = await service.getClientById(req.params.userId);
            res.status(DTO.statusCode).json(DTO);
        } catch (error) {
            next(error);
        }
    },
    updateClient: async (req, res, next) => {
        try {
            const DTO = await service.updateClient(req.params.userId, req.body);
            res.status(DTO.statusCode).json(DTO);
        } catch (error) {
            next(error);
        }
    },
    deleteAddressAdmin: async (req, res, next) => {
        try {
            const DTO = await service.deleteAddress(
                req.params.userId,
                req.query.addressId,
            );
            res.status(DTO.statusCode).json(DTO);
        } catch (error) {
            next(error);
        }
    },
    getAccounts: async (req, res, next) => {
        try {
            const DTO = await service.getAccounts(
                req.query.page,
                req.query.keyword,
            );
            res.status(DTO.statusCode).json(DTO);
        } catch (error) {
            next(error);
        }
    },
    updateAccount: async (req, res, next) => {
        try {
            const DTO = await service.updateAccount(
                req.user.role,
                req.user.uid,
                req.params.id,
                req.body,
            );
            res.status(DTO.statusCode).json(DTO);
        } catch (error) {
            next(error);
        }
    },
    resetPassword: async (req, res, next) => {
        try {
            const DTO = await service.resetPassword(
                req.user.role,
                req.params.id,
            );
            res.status(DTO.statusCode).json(DTO);
        } catch (error) {
            next(error);
        }
    },
    updateProfileClient: async (req, res, next) => {
        try {
            const DTO = await service.updateClient(req.user.uid, req.body);
            res.status(DTO.statusCode).json(DTO);
        } catch (error) {
            next(error);
        }
    },
    updateAddress: async (req, res, next) => {
        try {
            const DTO = await service.updateAddress(
                req.user.uid,
                req.params.addressId,
                req.body,
            );
            res.status(DTO.statusCode).json(DTO);
        } catch (error) {
            next(error);
        }
    },
    updateAddressAdmin: async (req, res, next) => {
        try {
            const DTO = await service.updateAddress(
                req.params.userId,
                req.query.addressId,
                req.body,
            );
            res.status(DTO.statusCode).json(DTO);
        } catch (error) {
            next(error);
        }
    },
    deleteAddressClient: async (req, res, next) => {
        try {
            const DTO = await service.deleteAddress(
                req.user.uid,
                req.params.addressId,
            );
            res.status(DTO.statusCode).json(DTO);
        } catch (error) {
            next(error);
        }
    },
    getAllOrders: async (req, res, next) => {
        try {
            const DTO = await service.getAllOrders(
                req.query.page,
                req.query.keyword,
            );
            res.status(DTO.statusCode).json(DTO);
        } catch (error) {
            next(error);
        }
    },
    getOrderByClientID: async (req, res, next) => {
        try {
            const DTO = await service.getOrderByClientID(req.user.uid);
            res.status(DTO.statusCode).json(DTO);
        } catch (error) {
            next(error);
        }
    },
    getOrderByIdClient: async (req, res, next) => {
        try {
            const DTO = await service.getOrderByID(
                req.params.orderId,
                req.user.uid,
            );
            res.status(DTO.statusCode).json(DTO);
        } catch (error) {
            next(error);
        }
    },
    getOrderByIdAdmin: async (req, res, next) => {
        try {
            const DTO = await service.getOrderByID(
                req.query.orderId,
                req.params.userId,
            );
            res.status(DTO.statusCode).json(DTO);
        } catch (error) {
            next(error);
        }
    },
    createStaff: async (req, res, next) => {
        try {
            const DTO = await service.createStaff(req.user.role, req.body);
            res.status(DTO.statusCode).json(DTO);
        } catch (error) {
            next(error);
        }
    },
    getOrdersByStatus: async (req, res, next) => {
        try {
            const DTO = await service.getOrdersByStatus(
                req.query.status,
                req.query.page,
                req.query.keyword,
            );
            res.status(DTO.statusCode).json(DTO);
        } catch (error) {
            next(error);
        }
    },
    updateOrder: async (req, res, next) => {
        try {
            const DTO = await service.updateOrder(
                req.user.uid,
                req.params.orderId,
                req.body,
            );
            res.status(DTO.statusCode).json(DTO);
        } catch (error) {
            next(error);
        }
    },
    processOrders: async (req, res, next) => {
        try {
            const DTO = await service.processOrders(
                req.user.uid,
                req.body,
                parseInt(req.query.type),
            );
            res.status(DTO.statusCode).json(DTO);
        } catch (error) {
            next(error);
        }
    },
    updatePasswordStaff: async (req, res, next) => {
        try {
            const DTO = await service.updatePasswordStaff(
                req.user.uid,
                req.body,
            );
            res.status(DTO.statusCode).json(DTO);
        } catch (error) {
            next(error);
        }
    },
};
